### PR TITLE
Implement export as namespace from

### LINKED ISSUE #4813: Support ES `export * as ns from` form
The ES7 proposal is available at: https://github.com/leebyron/ecmascript-more-export-from

The additions include:

reexporting default:

``` ts
// proposed:
export v from "mod";

// symmetric to:
import v from "mod";
export {v};
```

reexporting as sub-module:

``` ts
// proposed:
export * as ns from "mod";

// symmetric to:
import * as ns from "mod";
export {ns};
```

also allowing combining 

``` ts
// proposed
export v, {x, y as w} from "mod"
// symmetric to
import v, {x, y as w} from "mod"
As well as

// proposed
export v, * as ns from "mod"
// symmetric to
import v, * as ns from "mod"
```

### PR DESCRIPTION
**Pull request by Kingwl**
Fixes (a part of) #4813

### CODE REVIEW

**Comment by ExE-Boss** [comment]
This&nbsp;only addresses&nbsp;half of&nbsp;#4813. The&nbsp;other half&nbsp;being [`export default from`](https://tc39.es/proposal-export-default-from), which&nbsp;is&nbsp;currently&nbsp;at&nbsp;**stage&nbsp;1**. (unless&nbsp;it&nbsp;got&nbsp;moved to&nbsp;a&nbsp;separate&nbsp;issue)

**Comment by Kingwl** [comment]
up🙆🏻‍♂️

**Comment by Kingwl** [comment]
Could we get this down in 3.8?🤔

**Comment by weswigham** [comment]
@kingwl can ya sync it with `master`?

**Comment by Kingwl** [comment]
up🙋🏻‍♂️⛷

**Comment by Kingwl** [comment]
@typescript-bot pack this.

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -56,9 +56,10 @@ namespace ts {
                 break;
             // 4. Export alias declarations pointing at only uninstantiated modules or things uninstantiated modules contain
             case SyntaxKind.ExportDeclaration:
-                if (!(node as ExportDeclaration).moduleSpecifier && !!(node as ExportDeclaration).exportClause) {
+                const exportDeclaration = (node as ExportDeclaration);
```
Comment (saschanaz) [inline src/compiler/binder.ts]: Redundant parentheses?

**Comment by DanielRosenwasser** [comment]
@Kingwl I updated your branch for you, hope you don't mind!

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -36331,6 +36351,7 @@ namespace ts {
             case SyntaxKind.ImportClause: // For default import
             case SyntaxKind.ImportEqualsDeclaration:
             case SyntaxKind.NamespaceImport:
+            case SyntaxKind.NamespaceExport:
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: I don't think this is correct - namespace exports don't import anything locally.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -6623,6 +6638,7 @@ namespace ts {
                     case SyntaxKind.ImportSpecifier:
                     case SyntaxKind.NamedImports:
                     case SyntaxKind.NamespaceImport:
+                    case SyntaxKind.NamespaceExport:
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Same issue around namespace exports not being locally visible - I don't think any code should need to hit this path.

File: src/compiler/transformers/module/system.ts [hunk 4]
```diff
@@ -609,6 +633,13 @@ namespace ts {
             return singleOrMany(statements);
         }
 
+        function visitExportDeclaration(node: ExportDeclaration): VisitResult<Statement> {
+            if (node.exportClause && isNamespaceExport(node.exportClause)) {
+                hoistVariableDeclaration(getLocalNameForExternalImport(node, currentSourceFile)!); // TODO: GH#18217
```
Comment (DanielRosenwasser) [inline src/compiler/transformers/module/system.ts]: Does this need the TODO?

File: src/services/organizeImports.ts [hunk 5]
```diff
@@ -307,7 +307,7 @@ namespace ts.OrganizeImports {
         }
 
         const newExportSpecifiers: ExportSpecifier[] = [];
-        newExportSpecifiers.push(...flatMap(namedExports, i => (i.exportClause!).elements));
+        newExportSpecifiers.push(...flatMap(namedExports, i => i.exportClause && isNamedExports(i.exportClause) ? i.exportClause.elements : []));
```
Comment (DanielRosenwasser) [inline src/services/organizeImports.ts]: There's an `emptyArray` global you can use to avoid allocations.

**Comment by rbuckton** [comment]
@Kingwl: You need to modify the test for the visitor to `ExportDeclaration.exportClause` in `visitorPublic.ts`: https://github.com/microsoft/TypeScript/pull/34903/files?file-filters%5B%5D=.ts#diff-f3b57b1df462675a8d7ce1a0a1ef24f9R824

It should be switched from `isNamedExports` to `isNamedExportBindings` to match the new syntax.

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

**Comment by DanielRosenwasser** [comment]
I merged from master to fix conflicts, and I switched some of your tests to target multiple `module` outputs to avoid duplication (side note: I ran into #35809 while doing this which was annoying).

One thing I noticed was that you need a transform for `es2015`.

**Comment by weswigham** [comment]
@typescript-bot pack this again now that it's up to date

File: tests/baselines/reference/exportAsNamespace1(module=es2015).js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import * as foo from './1'
+
+foo.ns.a;
+foo.ns.b;
+
+//// [0.js]
+export var a = 1;
+export var b = 2;
+//// [1.js]
+export * as ns from './0';
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/exportAsNamespace1(module=es2015).js]: This needs to be transformed into something like

```js
import * as _ns from './0';
export { _ns as ns };
```

**Comment by DanielRosenwasser** [comment]
I implemented `export * as ns` in ES2015, and renamed the transform and file appropriately. I think we should be good to go! :rocket:

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

**Comment by DanielRosenwasser** [comment]
Thank you!!

**Pull request merged**